vmdk_plugin/drivers/photon/photon_driver.go

```diff
@@ -515,73 +583,45 @@ func (d *VolumeDriver) Path(r volume.Request) volume.Response {
-	// and we will happily decrement the refcount there, and will fail the unmount
-	// since the volume will have been never mounted.
-	// Note: for new keys, GO maps return zero value, so no need for if_exists.
-
-	refcnt := d.incrRefCount(r.Name) // save map traversal
-	log.Debugf("volume name=%s refcnt=%d", r.Name, refcnt)
-	if refcnt > 1 {
-		log.WithFields(
-			log.Fields{"name": r.Name, "refcount": refcnt},
-		).Info("Already mounted, skipping mount. ")
-		return volume.Response{Mountpoint: d.getMountPoint(r.Name)}
-	}
+	// lock the state
+	d.refCounts.StateMtx.Lock()
 
-	// This is the first time we are asked to mount the volume. The disk
-	// ID isn't always cached by the plugin on a host, the disk could have
-	// been created on a different host and getting mounted on another host.
-	// So on the first mount fetch the disk from Photon and add it's ID to
-	// the ref count map.
-	status, err := d.GetVolume(r.Name)
-	if err != nil {
-		d.decrRefCount(r.Name)
-		return volume.Response{Err: err.Error()}
-	}
-
-	fstype, exists := status[fsTypeTag]
-	if !exists {
-		fstype = fs.FstypeDefault
-	}
-
-	skipAttach := false
-	// If the volume is already attached to the VM, skip the attach.
-	if state, stateExists := status["State"]; stateExists {
-		if strings.Compare(state.(string), "DETACHED") != 0 {
-			skipAttach = true
-		}
-		log.WithFields(
-			log.Fields{"name": r.Name, "skipAttach": skipAttach},
-		).Info("Attached state ")
-	}
+	if d.refCounts.GetSuccess() != true {
+		// complete the mount process under the lock to avoid races
+		// dont do refcounting. Just mount
+		reponse := d.processMount(r, false)
+		d.refCounts.MarkDirty()
 
-	// Mount the volume and for now its always read-write.
-	mountpoint, err := d.MountVolume(r.Name, fstype.(string), status["ID"].(string), false, skipAttach)
-	if err != nil {
-		log.WithFields(
-			log.Fields{"name": r.Name, "error": err.Error()},
-		).Error("Failed to mount ")
-		d.decrRefCount(r.Name)
-		return volume.Response{Err: err.Error()}
+		// then unlock the state
+		d.refCounts.StateMtx.Unlock()
+		return reponse
 	}
 
-	return volume.Response{Mountpoint: mountpoint}
+	// if refcount has been succcessful,
+	// unlock and do usual normal mounting with refcounting as well
+	d.refCounts.StateMtx.Unlock()
+	return d.processMount(r, true)
 }
 
 // Unmount request from Docker. If mount refcount is drop to 0,
 // Unmount and detach from VM
 func (d *VolumeDriver) Unmount(r volume.UnmountRequest) volume.Response {
 	log.WithFields(log.Fields{"name": r.Name}).Info("Unmounting Volume ")
 
-	// dirty the background refcounting process if any
-	d.refCounts.MarkDirty()
+	// lock the state
+	d.refCounts.StateMtx.Lock()
+
+	if d.refCounts.GetSuccess() != true {
+		// if refcounting hasn't been succesful,
+		// no refcounting, no unmount. All unmounts are delayed
+		// until we succesfully populate the refcount map
+		d.refCounts.MarkDirty()
+		d.refCounts.StateMtx.Unlock()
```

Comment:
MarkDirty will just set the bool to true. It doesn't take any lock on its own. Just doing it as an exported method